Enable Highlight action on single word selection

poire-z opened the pull request:
Also add "Fulltext search" as an option for highlight action.
See https://github.com/koreader/koreader/issues/6111#issuecomment-623088751. Closes #6111.

@protist, in this [highlight_search_singleword.zip](https://github.com/koreader/koreader/files/4571118/highlight_search_singleword.zip) are the 2 modified files, if you want to confirm this works for you.

<kbd>![image](https://user-images.githubusercontent.com/24273478/80921250-84631f00-8d75-11ea-86fb-8d4dc0f1a6dc.png)</kbd>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/6114)
<!-- Reviewable:end -->

poire-z commented:
Dunno if we should also add `Dictionary lookup`.
It might work well only on single word selection - although I dunno if dicts can/do have multiwords word definitions and if sdcv works with that well.
But for a user that checks "Enable on single word selection", it can be handy I guess.

Frenzie commented:
@poire-z Works fine afaict ;-)

![Screenshot_20200503_212113](https://user-images.githubusercontent.com/202757/80923435-1c680500-8d84-11ea-9303-4286b44a206a.png)

mergen3107 commented:
@poire-z
Isn’t single word dictionary lookup currently set as default? On 2020.04.02 release at least. That’s what I extensively use right now

poire-z commented:
> Isn’t single word dictionary lookup currently set as default?

Yes it is. But people enabling the new option "Enable (highlight action) on single word selection" to be able to use one of the highlight action wouldn't have it in the set. So my question about including it.

These highlight actions can also be cycled with a gesture, so I guess some people might stick with it, and would cycle depending on the book content, and the need for quick "translate" or "wikipedia", or the good usual of "dictionary".

poire-z commented:
OK, this is buggy: when enabling "on single word", selecting multiple words only uses the first :)

poire-z commented:
OK, added "Dictionary", fixed bug.
Should we change the order, in the menu and in the cycling?
I did not, to not mess with current users that may be used to the current order.
<kbd>![image](https://user-images.githubusercontent.com/24273478/80927404-f3ed0480-8d9d-11ea-86fa-aab7393a237f.png)</kbd>

NiLuJe commented:
> Should we change the order, in the menu and in the cycling?

Current order looks fine to me ;).

protist commented:
Thank you so much for the quick work! This seems to work well for me. I had a few comments regarding the implementation though, back in [#6111](https://github.com/koreader/koreader/issues/6111#issuecomment-623239467).

poire-z commented:
https://github.com/koreader/koreader/issues/6111#issuecomment-623392959

|What|short hold|long hold at end of selection (3s)|
|----|----|----|
|Single word lookup |dict | button menu|
|Single word lookup with [x] Enable on single word lookup|highlight action|button menu|
|Multiple words selection|highlight action|button menu|

@protist: here's the 3rd version: [highlight_search_singleword3.zip](https://github.com/koreader/koreader/files/4574680/highlight_search_singleword3.zip)

protist commented:
@poire-z Looks great! This seems to work really well for my device! Thank you!

The last thing is I find the UX quite confusing for the settings. In particular, the `Enable on single word lookup` phrasing, but also the implied actions for the other inputs. I wonder if it would be possible to configure it similar to the gesture manager. That is, have a first tier where you specify the input trigger, i.e. select one of

* `Single word lookup — short hold`
* `Single word lookup — long hold`
* `Multiple words selection — short hold`
* `Multiple words selection — long hold`

Then a second tier where you specify one of the six associated actions, i.e. the popup dialog, highlight, translate, Wikipedia, dictionary, fulltext search.

poire-z commented:
Thanks for the feedback.

> I wonder if it would be possible to configure it similar to the gesture manager

Could make sense. But as it may have less simple implications to other code, I let that for some (someone else's :) other PR.

protist commented:
Thanks again @poire-z. That definitely sounds reasonable.

(Just for the record, ideally I think the multiple word selection would be different. e.g. if single-word lookup, short hold, is search, it doesn't make as much sense to force multiple-word selection, short hold, to be search too, but perhaps something like dictionary instead. Hence the finer tuning would be useful.)

poire-z commented:
(Just for the record too :)
I think the contrary: there's no real reason to have single word vs multiple words work differently - moreover with search: what you want to search for (or wikipedia lookup) might be one word, or two words: no reason for the user to wonder and think about what will happening.
It only makes sense with dictionary lookup, where we assume(d) dictionary have only definitions for single words.

protist commented:
Ooops, sorry, I mistyped. In my usage, probably 90% of the time I select multiple words it's for a highlight. For me, single words are never highlights, and are either search or dictionary.
